configuration/configuration.ts

```diff
@@ -0,0 +1,88 @@
+import {IConfiguration} from "./iconfiguration";
+import {FileConfiguration} from "./file-configuration";
+import {EnvConfiguration} from "./env-configuration";
+import {MongoConfiguration} from "./mongo-configuration";
+
+export class Configuration implements IConfiguration {
+    private static configFilename: string = "user-config.json";
+    private static mongoUriKey: string = "mongoUri";
+    private providers: IConfiguration[] = [];
+
+    /**
+     * Asynchronously initialize configuration providers.
+     *
+     * Configuration providers include, in order of preference:
+     *      - JSON file of user preference key-value pairs stored
+     *        in ./user-config.json
+     *      - Environment variables
+     *      - Values stored in a MongoDB collection
+     *
+     * Notes:
+     *      - The utilized Mongo collection must contain only a single document.
+     *      - The `mongoUri` key must be given a value before initialization.
+     *
+     * @param {string} collectionName - MongoDB collection to use for
+     * key-value storage.
+     * @param {string[]} requiredKeys - Keys required to be present in
+     * the MongoDB collection.
+     * @param {number} secondsToRetry - Time to wait for Mongo database to
+     * come online and contain requiredKeys, throwing an error after the
+     * allotted time.
+     */
+    public async initialize(
+            collectionName: string = "config",
+            requiredKeys: string[] = [],
+            secondsToRetry: number = 5 * 60): Promise<void> {
+        let fileConfig = new FileConfiguration();
+        let envConfig = new EnvConfiguration();
+        let mongoConfig = new MongoConfiguration();
+
+        await fileConfig.initialize(Configuration.configFilename);
+        this.providers = [fileConfig, envConfig];
+
+        let mongoUri: string = this.getString(Configuration.mongoUriKey);

```

Comment:
config keys might be split across all three providers indiscriminately: you shouldn't expect all requiredKeys to exist in mongo alone.
At this point, I suggest you remove all items in the array that exist in file/env, and just pass in the remaining items to mongoConfig. if MongoUri doesn't exist, throw an error.